overlord/devicestate/devicestate.go

```diff
@@ -475,45 +476,73 @@ func (ms *modelSnapsForRemodel) canHaveUC18PinnedTrack() bool {
 }
 
 type remodelVariant struct {
-	localSnapsRequired bool
+	offline        bool
+	localSnaps     []*snap.SideInfo
+	localSnapPaths []string
 }
 
 type pathSideInfo struct {
 	localSi *snap.SideInfo
 	path    string
 }
 
-func (ro *remodelVariant) UpdateWithDeviceContext(st *state.State,
-	pathSI *pathSideInfo, snapName string, opts *snapstate.RevisionOptions,
+func (ro *remodelVariant) UpdateWithDeviceContext(st *state.State, snapName string, snapID string, opts *snapstate.RevisionOptions,
 	userID int, snapStateFlags snapstate.Flags, tracker snapstate.PrereqTracker,
 	deviceCtx snapstate.DeviceContext, fromChange string) (*state.TaskSet, error) {
 	logger.Debugf("snap %s track changed", snapName)
 	if opts == nil {
 		opts = &snapstate.RevisionOptions{}
 	}
-	if ro.localSnapsRequired {
-		if pathSI == nil {
-			return nil, fmt.Errorf("no snap file provided for %q (track changed)", snapName)
+	if ro.offline {
+		pathSI := sideInfoAndPathFromID(ro.localSnaps, ro.localSnapPaths, snapID)
+		if pathSI != nil {
+			return snapstateUpdatePathWithDeviceContext(st, pathSI.localSi, pathSI.path, snapName, opts,
+				userID, snapStateFlags, tracker, deviceCtx, fromChange)
 		}
-		return snapstateUpdatePathWithDeviceContext(st, pathSI.localSi, pathSI.path, snapName, opts,
-			userID, snapStateFlags, tracker, deviceCtx, fromChange)
+
+		info, err := snapstate.CurrentInfo(st, snapName)
+		if err != nil {
+			if errors.Is(err, &snap.NotInstalledError{}) {
+				return nil, fmt.Errorf("no snap file provided for %q", snapName)
```

Comment:
we don't seem to reach here in tests

Reply:
Since `UpdateWithDeviceContext` is only called in cases where the target snap is already installed, this would be an unexpected case. Do you think I should remove the check for a `NotInstalledError` error, and add a comment about how this is unexpected?

Reply:
then maybe prefix if internal error: ... ?